frappe/core/doctype/user_permission/user_permission_list.js

```diff
@@ -1,22 +1,120 @@
 frappe.listview_settings['User Permission'] = {
+
 	onload: function(list_view) {
-		list_view.page.add_menu_item(__("Clear User Permissions"), () => {
+		var me = this;
+		list_view.page.add_inner_button( __("Add / Update"), function() {
+			let dialog =new frappe.ui.Dialog({
+				title : __('Add User Permissions'),
+				fields: [
+					{
+						fieldname: 'user',
+						label: __('For User'),
+						fieldtype: 'Link',
+						options: 'User',
+						reqd: 1,
+						onchange: function() {
+							dialog.fields_dict.doctype.set_input(undefined);
+							dialog.fields_dict.docname.set_input(undefined);
+							dialog.set_df_property("docname", "hidden", 1);
+							dialog.set_df_property("apply_to_all_doctypes", "hidden", 1);
+							dialog.set_df_property("applicable_doctypes", "hidden", 1);
+						}
+					},
+					{
+						fieldname: 'doctype',
+						label: __('Document Type'),
+						fieldtype: 'Link',
+						options: 'DocType',
+						reqd: 1,
+						onchange: function() {
+							me.on_doctype_change(dialog);
+						}
+					},
+					{
+						fieldname: 'docname',
+						label: __('Document Name'),
+						fieldtype: 'Dynamic Link',
+						options: 'doctype',
+						hidden: 1,
+						onchange:  function() {
+							let field = dialog.fields_dict["docname"];
+							if(field.value != field.last_value) {
+								if(dialog.fields_dict.doctype.value && dialog.fields_dict.docname.value && dialog.fields_dict.user.value){
+									me.get_applicable_doctype(dialog).then(applicable => {
+										me.get_multi_select_options(dialog, applicable).then(options => {
+											me.on_docname_change(dialog, options, applicable);
+										});
+									});
+								}
+							}
+						}
+					},
+					{
+						fieldname: 'apply_to_all_doctypes',
+						label: __('Apply to all Documents Type'),
+						fieldtype: 'Check',
+						checked: 1,
+						hidden: 1,
+						onchange: function() {
```

Comment:
rewrite this logic! on submit, it should query for all checked.

You are creating a redundant array of checked items.

Reply:
removed, now its is fetching only once